src/tyr/route_serializer_osrm.cc

```diff
@@ -1598,10 +1547,273 @@ json::ArrayPtr serialize_legs(const google::protobuf::RepeatedPtrField<valhalla:
+        for (size_t k = 0; k < route_leg_segs_by_dist[i][j].size(); k++) {
+          cache[i][j].emplace_back(std::string{});
+        }
+      }
+    }
+  }
+
+  ~route_summary_cache() {
+  }
+
+  size_t num_named_segments_for_route_leg(size_t route_idx, size_t leg_idx) {
+    if (route_idx >= route_leg_segs_by_dist.size()) {
+      return 0;
+    }
+
+    if (leg_idx >= route_leg_segs_by_dist[route_idx].size()) {
+      return 0;
+    }
+
+    return route_leg_segs_by_dist[route_idx][leg_idx].size();
+  }
+
+  // Return the n-part named-segment summary for the given route/leg.
+  // The summary returned is guaranteed to be comprised of as few named
+  // segments as possible, while also being unique among all route/same-leg
+  // summaries.
+  std::string get_n_segment_summary(size_t route_idx, size_t leg_idx, size_t num_named_segs) {
+    if (route_idx >= cache.size()) {
+      return "";
+    }
+
+    if (leg_idx >= cache[route_idx].size()) {
+      return "";
+    }
+
+    if (num_named_segs == 0) {
+      return "";
+    }
+
+    // num_named_segs is the number of named segments you'd like the summary to
+    // be comprised of. The smallest requestable value is 1 - so we store all
+    // summaries offset by n = num_named_segs - 1.
+    size_t n = num_named_segs - 1;
+    if (n >= cache[route_idx][leg_idx].size()) {
+      return "";
+    }
+
+    // empty summary means cache miss
+    if (cache[route_idx][leg_idx][n].empty()) {
+      // go figure out the summary
+      std::vector<const NamedSegment*> segs_by_maneuver_index;
+      segs_by_maneuver_index.reserve(num_named_segs);
+      for (size_t i = 0; i < num_named_segs; i++) {
+        segs_by_maneuver_index.emplace_back(&route_leg_segs_by_dist[route_idx][leg_idx][i]);
+      }
+
+      // sort by maneuver index
+      std::sort(segs_by_maneuver_index.begin(), segs_by_maneuver_index.end(),
+                [](const NamedSegment* a, const NamedSegment* b) { return a->index < b->index; });
+
+      std::string summary;
+      for (size_t i = 0; i < num_named_segs; i++) {
+        summary += segs_by_maneuver_index[i]->name;
+        if (i != num_named_segs - 1)
+          summary += ", ";
+      }
+
+      misses++;
+
+      cache[route_idx][leg_idx][n] = summary;
+    } else {
+      hits++;
+    }
+
+    return cache[route_idx][leg_idx][n];
+  }
+};
+
+std::vector<std::vector<std::string>>
+summarize_route_legs(const google::protobuf::RepeatedPtrField<DirectionsRoute>& routes) {
+
+  route_summary_cache rscache(routes);
+
+  // vector 1: routes
+  // vector 2: legs
+  // string: unique summary for the route/leg
+  std::vector<std::vector<std::string>> all_summaries;
+  all_summaries.reserve(routes.size());
+
+  // Find the simplest summary for every leg of every route. Important note:
+  // each route should have the same number of legs. Hence, we only need to make
+  // unique the same leg (leg_idx) between all routes.
```

Comment:
One (admittedly solvable) challenge with having leg-iteration as the outer loop is you don't know how many legs there are until you're on the route.  Otherwise, I don't see an algorithmic advantage, but maybe I'm missing it.  Keep in mind that even though the algorithm is O(n^2) the n's are small (n==number of routes) - and any route summary that is generated is cached, avoiding repeat "heavy" lifting (not really _that_ heavy).

Reply:
> the n's are small (n==number of routes) 
definitely makes sense 👍